### PR TITLE
v3: Proposal for advertisement.

### PR DESCRIPTION
**Pull request by Raynos**
@jcorbin put out an advertisement demo last week. 

We want to add advertisement to tchannel as a first class citizen
to remove the registration logic from autobahn.

I put out a seperate PR to move registration out of autobahn
into tchannel as a seperate serviceName with well defined
call requests and call responses ( #434 ).

Ideally we want the advertisement logic to grow up and become
first class frames and part of the protocol.

Here is an initial draft of what that might look like.

r: @jcorbin @kriskowal @mranney

### CODE REVIEW

File: docs/advertisement.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+server Y. It will advertise what services it has available
+at what costs.
+
+## Frames
+
+### Payload: advertise (0x20)
+
+Schema:
+```
+tracing:25 num:2 ( service~1 cost:1 ){num}
```
Comment (jcorbin) [inline docs/advertisement.md]: don't need tracing

File: docs/advertisement.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  - send a `forward-advertise` message back to the server that
+    advertised.
+
+It is valid to `advertise` multiple times down the same connection
+
+### Payload: relay-advertise (0x21)
+
+Schema:
+```
+tracing:25 num:2 ( service~1 hostport~1 cost:1 ){num}
```
Comment (jcorbin) [inline docs/advertisement.md]: don't need tracing

File: docs/advertisement.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+This means that if the tchannel server that receives the 
+`relay-advertise` message were to connect directly to the
+hostport it would be able to send call requests for the service
+and know that it gets delivered with the cost.
+
+### Payload: forward-advertise (0x22)
+
+Schema:
+```
+tracing:25 conncount:2
```
Comment (jcorbin) [inline docs/advertisement.md]: don't need tracing

File: docs/advertisement.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+A `forward-advertise` message is a response to an `advertise`
+message. It means that the remote intents to forward messages
```
Comment (jcorbin) [inline docs/advertisement.md]: s/intents/intends/

**Comment by jcorbin** [comment]
At least the frame specs should go into the protocol document.

Until we start breaking out longer feature-deep-dives into separate documents (such as fragmentation/streaming) we should probably just add an advertisement section to it.

**Comment by mranney** [comment]
I like advertise going into the protocol, but I'm worried about breaking existing implementations. Is there some way to do this with v2, or does it need to wait until v3?

Also, I think we should have the client regularly re-advertise on a server-specified interval. This should be something like 10 or 60 seconds. Upon receipt of an advertise message, a receiving router will fan this out to the other nodes in the egress affinity set. That way, a transient failure of a router node will be easy to recover form.

**Comment by jcorbin** [comment]
At this point we'll almost certainly punt any advertisement frame(s) until
v3.

I'd especially like to have a discussion about the beacon interval
semantics.  We're for sure doing them in v2, but I strongly suspect that we
simply don't need them, and moving to low level frames could be an
opportunity to get away from fixed beacon intervals.

**Comment by Raynos** [comment]
This will be completely punted until v3

**Pull request closed without merging**